File: js/tests/integration.test.ts
// tests/integration.test.ts

import style, { Color, Style } from "../src/index";

describe("Integration Tests: Styled Terminal Output", () => {
  let consoleLogSpy: jest.SpyInstance;

  beforeEach(() => {
    consoleLogSpy = jest.spyOn(console, "log").mockImplementation(() => {});
  });

  afterEach(() => {
    consoleLogSpy.mockRestore();
  });

  it("should output styled text to console.log with default style instance", () => {
    const styledText = style.bold.fg(Color.red).apply("Styled Text");
    console.log(styledText);
    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[1m\x1b[38;5;1mStyled Text\x1b[0m");
  });

  it("should output styled text to console.log with custom style instance", () => {
    const customStyle = new Style().underline.bg(Color.blue);
    const styledText = customStyle.apply("Another Styled Text");
    console.log(styledText);
    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[4m\x1b[48;5;4mAnother Styled Text\x1b[0m");
  });

  it("should handle complex style combinations and output correctly", () => {
    const complexStyle = style.bold.italic.underline
      .fg(Color.hex("#FF5733"))
      .bg(Color.hsl(240, 100, 50));
    const styledText = complexStyle.apply("Complex Style Test");
    console.log(styledText);

    const expectedColor = Color.hex("#FF5733").code;
    const expectedBgColor = Color.hsl(240, 100, 50).code;

    expect(consoleLogSpy).toHaveBeenCalledWith(
      `\x1b[1m\x1b[3m\x1b[4m\x1b[38;${expectedColor}m\x1b[48;${expectedBgColor}mComplex Style Test\x1b[0m`
    );
  });

  it("should handle reset styles correctly", () => {
    const resetStyle = style.bold.fg(Color.green).reset;
    const styledText = resetStyle.apply("Reset Text");
    console.log(styledText);
    expect(consoleLogSpy).toHaveBeenCalledWith("Reset Text");
  });

  it("should handle hidden and reveal styles correctly", () => {
    const hiddenStyle = style.hidden;
    const revealedStyle = style.reveal;

    console.log(hiddenStyle.apply("Hidden Text"));
    console.log(revealedStyle.apply("Revealed Text"));

    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[8mHidden Text\x1b[0m");
    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[28mRevealed Text\x1b[0m");
  });

  it("should handle light, normal, strikethrough, and noStrikethrough styles", () => {
    console.log(style.light.apply("Light Text"));
    console.log(style.normal.apply("Normal Text"));
    console.log(style.strikethrough.apply("Strikethrough Text"));
    console.log(style.noStrikethrough.apply("No Strikethrough Text"));

    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[2mLight Text\x1b[0m");
    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[22mNormal Text\x1b[0m");
    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[9mStrikethrough Text\x1b[0m");
    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[29mNo Strikethrough Text\x1b[0m");
  });

  it("should handle resetFg and resetBg styles", () => {
    console.log(style.fg(Color.red).resetFg.apply("Reset Foreground"));
    console.log(style.bg(Color.blue).resetBg.apply("Reset Background"));

    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[38;5;1m\x1b[39mReset Foreground\x1b[0m");
    expect(consoleLogSpy).toHaveBeenCalledWith("\x1b[48;5;4m\x1b[49mReset Background\x1b[0m");
  });
});

File: src/py/tests/integration.test.py
from styled_terminal import Style, Color, InvalidParameter
from unittest.mock import patch
import unittest


class TestStyledTerminalOutput(unittest.TestCase):

    def test_reset_style(self):
        # Applying the style with reset
        reset_style = Style().bold.fg(Color.green()).reset
        styled_text = reset_style.apply("Reset Text")

        # Assert that the styled text matches the expected output
        self.assertEqual(styled_text, "Reset Text")

    def test_strikethrough_styles(self):
        # Apply various styles
        light_text = Style().light.apply("Light Text")
        normal_text = Style().normal.apply("Normal Text")
        strikethrough_text = Style().strikethrough.apply("Strikethrough Text")
        no_strikethrough_text = Style().noStrikethrough.apply("No Strikethrough Text")

        # Assert that the output matches the expected ANSI escape sequences
        self.assertEqual(light_text, "\x1b[2mLight Text\x1b[0m")
        self.assertEqual(normal_text, "\x1b[21m\x1b[22mNormal Text\x1b[0m")
        self.assertEqual(strikethrough_text, "\x1b[9mStrikethrough Text\x1b[0m")
        self.assertEqual(no_strikethrough_text, "\x1b[29mNo Strikethrough Text\x1b[0m")

    def test_complex_style(self):
        # Complex combination of styles
        complex_style = (
            Style()
            .bold.italic.underline.fg(Color.hex("#FF5733"))
            .bg(Color.hsl(240, 100, 50))
        )
        styled_text = complex_style.apply("Complex Style Test")

        # Get the expected color and background codes
        expected_color = Color.hex("#FF5733").code
        expected_bg_color = Color.hsl(240, 100, 50).code

        # Assert the styled text matches the expected output with the escape codes
        self.assertEqual(
            styled_text,
            f"\x1b[1m\x1b[3m\x1b[4m\x1b[38;{expected_color}m\x1b[48;{expected_bg_color}mComplex Style Test\x1b[0m",
        )


if __name__ == "__main__":
    unittest.main()
